File: Java/Eta/Core/src/main/java/com/refinitiv/eta/transport/SocketBuffer.java
package com.refinitiv.eta.transport;

import java.nio.ByteBuffer;

class SocketBuffer extends EtaNode
{
    class SlicesPool extends Pool
    {
        SlicesPool(Object o)
        {
            super(o);
        }

        int _available;
        int _totalSliceCount;

        @Override
        EtaNode poll()
        {
            TransportBufferImpl slice;
            if ((slice = (TransportBufferImpl)super.poll()) != null)
            {
                --_available;
            }
            else
            {
                // create a new slice
                slice = new TransportBufferImpl(this);
                slice._data = _dataBuffer.duplicate();
                ++_totalSliceCount;
            }
            return slice;
        }

        @Override
        void add(EtaNode node)
        {
            super.add(node);
            ++_available;

            if (_available == _totalSliceCount)
            {
                // the SocketBuffer may being recycled
                SocketBuffer buffer = (SocketBuffer)_poolOwner;
                Pool sbp = buffer._pool;
                if (!sbp._isSharedPoolBuffer)
                {
                    if (!sbp._isProtocolBuffer)
                    {
                        ((RsslSocketChannel)sbp._poolOwner).socketBufferToRecycle(buffer);
                    }
                    else
                    {
                        buffer.returnToPool();
                    }
                }
                else
                {
                    ((ServerImpl)sbp._poolOwner).socketBufferToRecycle(buffer);
                }
            }
        }

        boolean areAllSlicesBack()
        {
            return (_available == _totalSliceCount);
        }
    }

    static final int PACKED_HDR = 2;
    static final int RIPC_WRITE_POSITION = 3;
    static final byte FRAGMENT_RIPC_FLAGS = 0x04;
    static final byte FRAGMENT_HEADER_RIPC_FLAGS = 0x08;
    static final int COMPRESSION_EXTRA_LEN = 16;

    ByteBuffer _dataBuffer;

    // pool of slices
    final SlicesPool _slicesPool;

    int _bytesUsed = 0;
    boolean _isPacked = false;

    SocketBuffer(Pool pool, int size)
    {
        _dataBuffer = ByteBuffer.allocate(size);
        pool(pool);
        _slicesPool = new SlicesPool(this);
    }

    TransportBufferImpl getBufferSlice(int size, boolean packedBuffer)
    {
        // locked by calling method
        TransportBufferImpl slice = null;

        int _headerLength = RIPC_WRITE_POSITION;
        if (!packedBuffer)
        {
            _isPacked = false;
        }
        else
        {
            _headerLength += PACKED_HDR;
            _isPacked = true;
        }

        if ((_dataBuffer.capacity() - _bytesUsed) >= (_headerLength + size))
        {
            slice = (TransportBufferImpl)_slicesPool.poll(); // it returns non null
            slice._data.clear();

            slice._startPosition = _bytesUsed;
            slice._data.limit(_bytesUsed + _headerLength + size);
            slice._data.position(_bytesUsed + _headerLength);
            slice._length = size;
            _bytesUsed = _bytesUsed + _headerLength + size;

            // has to mark the position for the packed message length
            if (_isPacked)
            {
                slice._isPacked = true;
                slice._packedMsgLengthPosition = slice._startPosition + RIPC_WRITE_POSITION;
            }
            else
            {
                slice._isPacked = false;
            }
        }

        return slice;
    }

    TransportBufferImpl getBufferSliceForFragment(int size)
    {
        // locked by calling method
        TransportBufferImpl slice = null;
        _isPacked = false;

        slice = (TransportBufferImpl)_slicesPool.poll();
        slice._startPosition = 0;
        slice._data.limit(size);
        slice._data.position(0);
        _bytesUsed = size;

        return slice;
    }

    void clear()
    {
        _dataBuffer.clear();
        _bytesUsed = 0;
    }

}

File: CSharp/Eta/Src/Core/Transport/internal/SocketBuffer.cs
/*|-----------------------------------------------------------------------------
 *|            This source code is provided under the Apache 2.0 license
 *|  and is provided AS IS with no warranty or guarantee of fit for purpose.
 *|                See the project's LICENSE.md for details.
 *|           Copyright (C) 2022-2023 LSEG. All rights reserved.     
 *|-----------------------------------------------------------------------------
 */

using LSEG.Eta.Common;
using LSEG.Eta.Internal.Interfaces;
using LSEG.Eta.Transports;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace LSEG.Eta.Internal
{
    internal class SocketBuffer : EtaNode
    {
        public class SlicesPool : Pool
        {
            int m_Available = 0;
            int m_TotalSliceCount = 0;
            SocketBuffer m_SocketBuffer;

            public SlicesPool(Object owner) : base(owner)
            {
                m_SocketBuffer = owner as SocketBuffer;
            }

            public override EtaNode Poll()
            {
                TransportBuffer slice;
                if ((slice = (TransportBuffer)base.Poll()) != null)
                {
                    --m_Available;
                    slice.StartPosition = 0;
                    slice.HeaderLength = RIPC_WRITE_POSITION;
                }
                else
                {
                    ByteBuffer byteBuffer = new ByteBuffer(m_SocketBuffer.ByteBuffer, true);

                    // create a new slice
                    slice = new TransportBuffer(this, byteBuffer);
                    ++m_TotalSliceCount;
                }
                return slice;
            }

            public override void Add(EtaNode node)
            {
                base.Add(node);
                ++m_Available;

                if (m_Available == m_TotalSliceCount)
                {
                    // the SocketBuffer may being recycled
                    Pool socketBufferPool = m_SocketBuffer.Pool;
                    if (!socketBufferPool.IsSharedPoolBuffer)
                    {
                        if (!socketBufferPool.IsProtocolBuffer)
                        {
                            ((ChannelBase)socketBufferPool.PoolOwner).SocketBufferToRecycle(m_SocketBuffer);
                        }
                        else
                        {
                            m_SocketBuffer.ReturnToPool();
                        }

                    }
                    else
                    {
                        ((ServerImpl)socketBufferPool.PoolOwner).SocketBufferToRecycle(m_SocketBuffer);
                    }
                }
            }

            public bool AreAllSlicesBack()
            {
                return (m_Available == m_TotalSliceCount);
            }
        }

        internal const int PACKED_HDR = 2;
        internal const int RIPC_WRITE_POSITION = 3;
        internal const byte FRAGMENT_RIPC_FLAGS = 0x04;
        internal const byte FRAGMENT_HEADER_RIPC_FLAGS = 0x08;
        internal const int COMPRESSION_EXTRA_LEN = 16;

        internal byte[] ByteBuffer { get; private set; }
        public SlicesPool m_SlicesPool;
        int m_BytesUsed = 0;

        public SocketBuffer(Pool pool, int size)
        {
            ByteBuffer = new byte[size];
            Pool = pool;
            m_SlicesPool = new SlicesPool(this);
        }

        public TransportBuffer GetBufferSlice(int size, bool packedBuffer, int headerLength)
        {
            // locked by calling method
            TransportBuffer slice = null;
            bool isPacked;

            int _headerLength = headerLength;
            if (!packedBuffer)
            {
                isPacked = false;
            }
            else
            {
                _headerLength += PACKED_HDR;
                isPacked = true;
            }

            if ((ByteBuffer.Length - m_BytesUsed) >= (_headerLength + size))
            {
                slice = (TransportBuffer)m_SlicesPool.Poll(); // it returns non null
                slice.Data.Clear();

                slice.StartPosition = m_BytesUsed;
                slice.Data.Limit = m_BytesUsed + _headerLength + size;
                slice.Data.WritePosition = m_BytesUsed + _headerLength;
                m_BytesUsed = m_BytesUsed + _headerLength + size;

                // has to mark the position for the packed message length
                if (isPacked)
                {
                    slice.IsPacked = true;
                    slice.PackedMsgOffSetPosition = slice.StartPosition + headerLength;
                }
                else
                {
                    slice.IsPacked = false;
                }
            }

            return slice;
        }

        public TransportBuffer GetBufferSliceForFragment(int size)
        {
            // locked by calling method
            TransportBuffer slice;

            slice = (TransportBuffer)m_SlicesPool.Poll();
            slice.Data.Clear();
            slice.StartPosition = 0;
            slice.Data.Limit = size;
            m_BytesUsed = size;

            return slice;
        }

        public void Clear()
        {
            m_BytesUsed = 0;
        }
    }
}
